src/c_kzg_4844.c

```diff
@@ -2378,231 +2379,84 @@ typedef struct {
-        if (ret != C_KZG_OK) goto out;
-        for (size_t j = 0; j < len_out; j++) {
-            blst_fr_mul(&mul_eval_ps[j], &mul_eval_ps[j], &p_eval[j]);
+        for (size_t j = i - 1; j > 0; j--) {
+            blst_fr_mul(&poly[j], &poly[j], &neg_root);
+            blst_fr_add(&poly[j], &poly[j], &poly[j - 1]);
         }
+        blst_fr_mul(&poly[0], &poly[0], &neg_root);
     }
 
-    ret = ifft_fr(out->coeffs, mul_eval_ps, len_out, s);
-    if (ret != C_KZG_OK) goto out;
-
-    out->length = out_degree + 1;
+    poly[roots_len] = FR_ONE;
+    *poly_len = roots_len + 1;
 
-out:
-    return ret;
+    return C_KZG_OK;
 }
 
-/**
- * Calculate the minimal polynomial that evaluates to zero for the powers of
- * roots of unity that correspond to missing indices.
+/* Computes the minimal polynomial that evaluates to zero at equally
+ * spaced chosen roots of unity in the domain of size `FIELD_ELEMENTS_PER_BLOB`.
  *
- * This is done simply by multiplying together `(x - r^i)` for all the `i` that
- * are missing indices, using a combination of direct multiplication
- * (#do_zero_poly_mul_partial) and iterated multiplication via convolution
- * (#reduce_partials).
+ * The roots of unity are chosen based on the missing cell indices. If the i'th
+ * cell is missing, then the i'th root of unity will be zero on the polynomial
+ * computed, along with every `CELLS_PER_EXT_BLOB` spaced root of unity in the
+ * domain.
  *
- * @param[out]  zero_poly       The zero polynomial
- * @param[out]  zero_poly_len   The zero polynomial length
- * @param[in]   missing_indices The indices of the missing coefficients
- * @param[in]   len_missing     The number of missing indices
- * @param[in]   s               The trusted setup
+ * @param[in,out]   zero_poly      The zero polynomial for cell indices
+ * @param[in,out]   zero_poly_len  The length of zero_poly
+ * @param[in]       missing_cell_indices     The array of missing cell indices
+ * @param[in]       len_missing_cells The number of missing cell indices
+ * @param[in]       s         The trusted setup
  *
- * @remark This does not work if all indices are missing.
- * @remark Unused coefficients are set to zero.
+ * @remark This does not work if all cells are missing.
+ * @remark `missing_cell_indices` are assumed to be less than
+ * `CELLS_PER_EXT_BLOB`.
  */
-static C_KZG_RET zero_polynomial_via_multiplication(
+static C_KZG_RET zero_polynomial_periodic_roots(
     fr_t *zero_poly,
     size_t *zero_poly_len,
-    const uint64_t *missing_indices,
-    size_t len_missing,
+    const uint64_t *missing_cell_indices,
+    size_t len_missing_cells,
     const KZGSettings *s
 ) {
     C_KZG_RET ret;
-    fr_t *work = NULL;
-    fr_t *scratch = NULL;
-    poly_t *partials = NULL;
+    fr_t *roots = NULL;
+    fr_t *short_zero_poly = NULL;
+    size_t short_zero_poly_len = 0;
 
     /* If nothing is missing, return all zeros */
-    if (len_missing == 0) {
+    if (len_missing_cells == 0) {
```

Comment:
I am a bit puzzled about this special case (which was already around before this PR): the vanishing polynomial of empty set should be constant FR_ONE, right? Because it is defined as the polynomial that vanishes exactly on the set, or as the empty product over terms `X-r_i`.  

If it is a special case that we really need for the recovery, then we should document that. Are we expecting users to call recovery even no cell is missing? 

Otherwise, I would suggest we explicitly disallow this case, and return `C_KZG_BADARGS`, as you do in `zero_polynomial_from_roots`.

Reply:
returning `C_KZG_BADARGS` would be the cleanest way I think -- will change to that and add a remark that the user should explicitly handle that case.